### DISCUSSION TITLE
Module support for Tasmota

### DISCUSSION THREAD

**Discussion by gemu2015** [arendst/Tasmota#12070] [Ideas]
i want to introduce a new concept for Tasmota drivers: modules
a long time ago i worked several years with a real time multitasking operating system OS9 that was far ahead of its competitors.
i wrote a lot of drivers for that system all in assembly code at that time.
the drivers are relocatable and can be linked and unlinked at runtime, no reboot needed.

and i always wanted this driver concept for Tasmota.

i now have a first working version with a sample driver for Tasmota. (relocatable ESP8266 sensor driver)

The pros:
you may link every driver you want (add it to the standard Tasmota drivers, completely compatible with existing system)
you may use all Flash for the drivers (not only half of it)
you may exchange drivers at runtime

The cons:
drivers have to be adapted. also with mostly little effort. (MACROS provided)
the drivers may not use c++, only c. would require some drivers to be modified.
if you update the core binary, all drivers are lost and have to be reloaded. should be possible 
to automize in the future.
we would get a pool of binary drivers, each 2 versions for ESP8266 and ESP32
binaries are about 10% larger then before

new cmds: (inspired by OS9)

link /module.bin 
links a module from filesystem to the next free memory slot. will be also possible via upload in a future version

unlink X
unlinks (deletes) the module Nr x from system

iniz X
initializes the module Nr x  (attaches it to Tasmota)

deiniz X
deinitializes the module Nr x (detaches it from Tasmota and free memory)

mdir
show a directory of linked modules

dump X
shows a memory dump of module Nr x

i am not willing to adapt all drivers, only a few samples.
this concept only makes sense with a significant number of available drivers.
it would require that other coders probably would adapt the drivers they have written.
probably there are better macros possible to reduce the adaptation effort

currently the development system is a little laborious and in an early stage.
i will describe the procedure if someone is interested to participate.

**Comment by s-hadinger** [comment]
Cool. I’m interested but its unclear to me how this is compatible with esp-idf and Arduino. And how the tool chain can be simple enough.

**Reply by gemu2015** [reply]
it is completely compatible with existing tasmota and esp-idf and can be mixed with existing drivers

**Comment by gemu2015** [comment]
the toolchain is the problem. for several days i tried to handle the linker scripts and gcc c and assembler options and in the end i gave up and solved it my own way. the driver is compiled within the tasmota environment as part of the normal toolchain (only 1 driver at a time) then when testing all functionality is done  i do the trick and copy the driver part out of the binary into to file. since the driver is completely relocatable this works without problems. this file then can be loaded into other devices via file system or later via upload.

we will have to create a kind of batch that enables only 1 driver at a time, compiles it, then copies the driver part from the binary and then next one

**Reply by s-hadinger** [reply]
It raises many questions:
- how do you know which part belongs to the driver in the binary?
- How do you call back Tasmota functions? Do you plan to have a vtable for some Tasmota and Arduino functions?

**Comment by gemu2015** [comment]
i put the driver into special sections and in the adapted linker script places them in the right order, so no problem at all.
of course i need jump tables for all used functions of the driver. all is done via a master pointer which must be passed to every function that wants to use system calls.  that jump tables system calls may be expanded as needed time by time. the master pointer also has entries for tasmota settings and memory usage.  since there is no startup code the driver must request memory and initialize all its variables as needed. most is done via macros. this is all straight forward and works well so far.
the system and Tasmota calls get the character j before the function name  so jcalloc()  or jfree() it is quite easy to transform.
i looked at several drivers and the used system or Tasmota calls are foreseeable.

**Reply by s-hadinger** [reply]
Nice. Last question for now, do you need to load drivers into ram or you can run them from filesystem?

**Comment by gemu2015** [comment]
i am currently on ESP8266, no chance to load a driver into tiny executable RAM. on ESP32 this may work when you can allocate enough EXE RAM.
normally i load the drivers into flash (after the end of normal binary) by this you can use the rest of the flash that is normally not usable because of  the swapping during update. as already stated doing an update of main program will however overwrite the drivers and they have to be reloaded.

**Reply by gemu2015** [reply]
and yes you can load from filesystem to flash, that is what i am currently doing. since the 1 M devices dont have a file system we need a WEB uploader in the future

**Reply by barbudor** [reply]
I think @s-hadinger question was more about "execute in place" (XIP) as the filesystem may be already in flash and the module doesn't need to be relocated.

**Reply by gemu2015** [reply]
this would work if the flash file system puts the file in consecutive sectors. then if we could get the start of the file in flash the module could be executed directly. i am not aware of an api call to get this pointer.

**Reply by s-hadinger** [reply]
I believe there are some restrictions and all areas in memory are not necessarily executable. But there is also a mmu on esp32, maybe it can do interesting things

**Reply by gemu2015** [reply]
you can use all flash on ESP chips for execution. this gives twice the memory for executables if we accept that all modules are lost on reflashing at least with OTA updates. but since every driver can be replaced with new versions at any time we dont need updates of the core program that often.
